### PR TITLE
Add levelConforms feature

### LINKED ISSUE #15: Allow checking for level of feature and level validity
It came up in work on edtf-validate whether we want to say something is of a Level if it is a feature of that level specifically as described in the EDTF specification or if it is a date valid in that level (meaning something that is valid in level 0 must also be valid at level 1, since to support level 1 you must support level 0).

An example in question: `2013/2014` is a time interval described in Level 0, so  it validates at level 0 then, and is a feature of level 0, and since all of level 0 must be supported by a system that supports level 1, it also conforms to level 1.

Response from @vphill regarding how we want to handle this where it is used by our edtf service (django-edtf):

so this
https://digital2.library.unt.edu/edtf/isValid.json?date=1985
returns validEDTF: true

Asking for Level 0
https://digital2.library.unt.edu/edtf/isValid.json?date=1985&level=0
returns validEDTF: true

Asking for Level 1
https://digital2.library.unt.edu/edtf/isValid.json?date=1985&level=1
returns validEDTF: false

and asking for Level 2
https://digital2.library.unt.edu/edtf/isValid.json?date=1985&level=2
returns validEDTF: false

So it seems like the current implementation is more of "Is the date value a feature of level" instead of does it validate at this level

Could we keep the current behavior of level, possibly changing the argument to `levelFeature`

And then we could add a `levelConforms` which would be "does this date conform/validate at this given levelConforms"

So
 date=2003&levelConforms=0 would be True
 date=2003&levelConforms=1 would be True
 date=2003&levelConforms=2 would be True
Because Level 2 requires support for Level 1 and Level 0

I'm not married to the api names if there are ones you like better.  just tried to put something down

### PR DESCRIPTION
**Pull request by madhuBayy**
@ldko @somexpert 
This closes #15 by adding levelConfirms feature.

### CODE REVIEW

**Comment by madhuBayy** [comment]
@ldko @somexpert This is ready for review again now.

File: tox.ini [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
-    py{27, 36, 37}-django111
-    py37-flake8
+    py{27, 36, 37, 38}-django111
+    py38-flake8
 
 [flake8]
 max-line-length = 99
 
 [testenv]
+download = True
```
Comment (somexpert) [inline tox.ini:14]: Why is this needed?

File: tests/tests.py [hunk 2]
```diff
@@ -48,21 +48,43 @@ def test_returns_true_with_valid_date_no_level(self):
...
+    def test_returns_400_with_valid_date_and_invalid_levelConforms(self):
+        response = self.client.get(reverse('edtf:results_json'), {'date': '2013',
+                                                                  'levelConforms': '3'})
+        self.assertEqual(response.status_code, 400)
+
+    def test_returns_400_with_valid_date_and_empty_levelConforms(self):
+        response = self.client.get(reverse('edtf:results_json'), {'date': '2013',
+                                                                  'levelConforms': ''})
+        self.assertEqual(response.status_code, 400)
+
```
Comment (somexpert) [inline tests/tests.py]: I feel like we could use some additional tests. Maybe a test with a valid date against levelConforms 0? And how about some tests against levelConforms where the date is invalid?

**Review by somexpert — COMMENTED** [review]
Just a question and a couple recommendations for added tests.

File: tox.ini (hunk 1 above)
Comment (madhuBayy) [reply tox.ini:14]: The py36 travis build was failing initially. So I have added it to get the latest versions in the venv to fix it.

File: tests/tests.py (hunk 2 above)
Comment (madhuBayy) [reply tests/tests.py]: sure! I can add those. I think I can add some invalid dates to levelFeature url too.

File: edtf/views.py [hunk 3]
```diff
@@ -40,6 +40,19 @@ def result_json(request):
         jsonedtfdict = {
             'validEDTF': valid_edtf.is_valid(date) and funcs[level]
         }
+    elif 'levelConforms' in request.GET:
+        levelConforms = request.GET.get('levelConforms')
+        if levelConforms == '' or levelConforms is None or int(levelConforms) > 2 or \
+                int(levelConforms) < 0:
```
Comment (ldko) [inline edtf/views.py]: Since we only have three valid levels, what do you think about setting a constant of `LEVELS = ('0', '1', '2')` near the top of this file and using it for checking `level` and `levelConforms` in the conditional with something like:
```
if `level` not in LEVELS
```
and
```
if levelConforms not in LEVELS:
```

File: tests/tests.py (hunk 2 above)
Comment (ldko) [reply tests/tests.py]: Seems like some parametrization could be worthwhile, though I suppose it could get confusing...

**Review by somexpert — APPROVED** [review]
This looks good to me :thumbsup:

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,6 +7,8 @@ x.x.x
 
 * Update requirements.txt to get latest version of edtf-validate
 * Updated project to python 3.7 
+* Added LevelConforms feature introduced in edtf-validate.
```
Comment (ldko) [inline CHANGELOG.md]: I suggest making this a little more specific with something like "Added `levelConforms` query parameter to check `conformsLevel*` functionality now present in edtf-validate."

File: CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,8 @@ x.x.x
...
+* Renamed 'level' feature to 'levelFeature'.
```
Comment (ldko) [inline CHANGELOG.md]: I would use back ticks on here instead of single quotes.

**Review by ldko — COMMENTED** [review]
I noticed we need to update the README to update the EDTF specification link and to reflect Python 3.8 support. I think it would make sense to also mention about the `levelFeature` and `levelConforms` there.

In running the app, I also noticed that we need to update the [edtf_form template](https://github.com/unt-libraries/django-edtf/blob/master/edtf/templates/edtf/edtf_form.html) to reflect the API changes.

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,8 @@ x.x.x
 
 * Update requirements.txt to get latest version of edtf-validate
 * Updated project to python 3.7 
+* Added levelConforms query parameter to check `conformsLevel*` functionality now present in edtf-validate.
```
Comment (ldko) [inline CHANGELOG.md]: Would you mind putting `levelConforms` in back ticks?

Also, if you don't mind doing that, would you mind changing "Update requirements" a couple lines up to "Updated requirements" and add periods to the end of the statements in the changelog where they are missing--for consistency's sake? :pleading_face:

File: README.md [hunk 7]
```diff
@@ -7,18 +7,20 @@ About
 -----
 
 Django-edtf is an app which provides the service of validating dates that comply with the
-Extended Date Time Format ([Levels 0-2](http://www.loc.gov/standards/datetime/pre-submission.html)).
+Extended Date Time Format ([Levels 0-2](http://www.loc.gov/standards/datetime)).
```
Comment (ldko) [inline README.md]: I would use https (https://www.loc.gov/standards/datetime/).

File: edtf/templates/edtf/edtf_form.html [hunk 8]
```diff
@@ -43,11 +43,15 @@ <h1>
     "validEDTF": true,
 })
```
Comment (ldko) [inline edtf/templates/edtf/edtf_form.html:45]: There is another EDTF spec link in the intro above that should be updated to https://www.loc.gov/standards/datetime/.

File: edtf/templates/edtf/edtf_form.html [hunk 9]
```diff
@@ -70,7 +74,8 @@ <h1>
 <h4>Parameters</h4>
 <b>date</b>: <i>Required</i> - the text respresentation of the date you want to verify.<br>
 <b>callback</b>: <i>Optional</i> - the text represenation of the jsonp callback wrapper.<br>
-<b>level</b>: <i>Optional</i> - the level at which you would like the given date verified against. If none is given, the service assumes you are verifying against the full EDTF spec.
+<b>levelFeature</b>: <i>Optional</i> - the level at which you would like the given date should be a feature of. If none is given, the service assumes you are verifying against the full EDTF spec.<br>
```
Comment (ldko) [inline edtf/templates/edtf/edtf_form.html]: I think this wording needs to be tweaked sightly. Perhaps something like: "...the level at which you would like the given date checked for being a feature of..."

Comment (ldko) [reply edtf/templates/edtf/edtf_form.html]: Also, what do you think about removing "If none is given, the service assumes you are verifying against the full EDTF spec." from the `levelFeature` and `levelConfoms` lines here, since it only makes sense if both `levelFeature` and `levelConforms` are left off the request?

Comment (madhuBayy) [reply edtf/templates/edtf/edtf_form.html]: I think it's ok to let it be here cos, it mentions if the optional arguments `levelFeature` and `levelConforms` are not given, the date will then be validated against full edtf spec. I can have it removed if you think otherwise.

Comment (ldko) [reply edtf/templates/edtf/edtf_form.html]: The way I read it, if you don't have `levelFeature`, whether or not you have `levelConforms`, it would be validating against the full spec. Maybe "If none" should be changed to "If neither `levelFeature` nor `levelConforms`" and, in the next line, "If neither `levelConforms` nor `levelFeature`" Or perhaps the "If none" sentences should be removed and a "Note:" added at the end of the Parameters section that indicated `levelFeature` and `levelConforms` are mutually exclusive, and if neither is given, the `date` will be validated against the full EDTF specification.

File: edtf/templates/edtf/edtf_form.html [hunk 10]
```diff
@@ -38,16 +38,20 @@ <h1>
     "validEDTF": false,
 }
 
-$ curl "http://{{ request.get_host }}/edtf/isValid.json?date=1234&callback=foo"
+$ curl http://{{ request.get_host }}/edtf/isValid.json?date=1234&callback=foo
```
Comment (ldko) [inline edtf/templates/edtf/edtf_form.html]: I think we should bring back these quotes because otherwise the URL gets chopped at the `&` because it is a control character. We should add quotes around all the urls that have an ampersand.

Actually in trying this, regarding #12, I think we should also add `-L` to all of the `curl` commands so it will follow an http->https redirect or we need to maybe make a scheme setting.

File: edtf/templates/edtf/edtf_form.html [hunk 11]
```diff
@@ -21,7 +21,7 @@ <h1>
...
+To determine whether a date conforms to the <a href="https://www.loc.gov/standards/datetime/">Extended Date Time Format Level 0-2</a>, enter it into the field below and click "Submit". A level 0 EDTF string can be comprised of 1 of 3 features (Date, Date Time, or Interval). A level 1 EDTF string consists of the same features, but may incorportate uncertainty, approximations, and unspecified dates. Level 1 may also feature years with more than 4 digits, and/or seasons. Additionally, level 2 includes partial uncertain/approximate and partial unspecified dates. Level 2 also features multiple dates as lists, masked precision, extended intervals, exponential and scientific notation for 'big' years. The result of this form will be either 'Valid EDTF date' or 'Invalid EDTF date'.<br>
```
Comment (ldko) [inline edtf/templates/edtf/edtf_form.html]: Just noticed we should remove ", masked precision" from the line above as it is no longer an edtf thing.

**Review by ldko — APPROVED** [review]
This looks good to me! Thank you, @madhulika95b !
